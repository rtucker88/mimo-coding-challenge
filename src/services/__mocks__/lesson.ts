import { ILesson } from "../lesson";

export const lessons: ILesson[] = [{
    content: [{
        color: '#fff',
        text: 'hello'
    }],
    id: 0
},
{
    content: [{
        color: '#e4e4e4',
        text: 'world'
    }],
    id: 1
}];

export function getLessons(): Promise<ILesson[]> {
    return new Promise((resolve, reject) => {
        process.nextTick(() => {
            resolve(lessons);
        });
    });
}
