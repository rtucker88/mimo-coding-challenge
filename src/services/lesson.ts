export interface IContent {
    color: string;
    text: string;
  }
  
interface ILessonInput {
    startIndex: number;
    endIndex: number;
}

export interface ILesson {
    id: number;
    content: IContent[];
    input?: ILessonInput;
}

export function getLessons(): Promise<ILesson[]> {
    return fetch('https://file-bzxjxfhcyh.now.sh/')
      .then(res => res.json())
      .then(res => res.lessons);
}
