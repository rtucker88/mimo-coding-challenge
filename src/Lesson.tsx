import * as classnames from 'classnames';
import { flatten, isUndefined } from 'lodash';
import * as React from 'react';
import { Link } from 'react-router-dom';
import { StateHandler, StateHandlerMap, withStateHandlers } from 'recompose';
import { ILesson } from './services/lesson';

import './lesson.css';

interface ILessonProps {
    lesson: ILesson;
    nextLesson: number | 'DONE';
    completeLesson: () => void;
}

interface ILessonState {
    isLessonFinished: boolean;
}

interface ILessonStateHandlers extends StateHandlerMap<ILessonState> {
    onInputChange: StateHandler<ILessonState>;
}

export const enhance = withStateHandlers<ILessonState, ILessonStateHandlers, ILessonProps>((props) => ({
    isLessonFinished: isUndefined(props.lesson.input),
}), {
    onInputChange: (_, props) => (event: React.ChangeEvent<HTMLInputElement>) => {
        return {
            isLessonFinished: getSolutionValue(props.lesson) === event.target.value
        };
    }
});

export type CombinedLessonProps = ILessonProps & ILessonState & ILessonStateHandlers;

const Lesson: React.StatelessComponent<CombinedLessonProps> = ({ lesson, nextLesson, completeLesson, onInputChange, isLessonFinished }) => {
    return (<div>
        {createLesson(lesson, onInputChange)}
        <Link to={`/${nextLesson}`} className={classnames({'disabled-link': !isLessonFinished})}><button disabled={!isLessonFinished} onClick={completeLesson}>Continue</button></Link>
    </div>);
};

export default enhance(Lesson);

export function getSolutionValue({ content, input }: ILesson): string {
    return content.map(cnt => cnt.text).join('').substr(input!.startIndex, input!.endIndex - input!.startIndex);
}

// Assuming valid data when going through this process. The error cases quickly result in a broken UI and 
// I'm unsure how we would recover from that state
function createLesson(lesson: ILesson, onInputChange: (event: React.ChangeEvent<HTMLInputElement>) => Partial<ILessonState> | undefined) {
    let textCounter = 0; // Use this to figure out where we replace text with an input
    
    const elems = lesson.content.map(entry => {
        const style: React.CSSProperties = {color: entry.color};
        const retElems: JSX.Element[] = [];

        const startIndex = textCounter;
        const endIndex = textCounter + entry.text.length;

        // Tokenize around the input
        if (lesson.input && lesson.input.startIndex >= startIndex && lesson.input.endIndex < endIndex) {
            // Before Input
            retElems.push(<span style={style} key={`${textCounter}-before`}>{entry.text.substr(0, lesson.input.startIndex - startIndex)}</span>)
            // Input
            retElems.push(<input style={style} key={textCounter} maxLength={lesson.input.endIndex - lesson.input.startIndex} onChange={onInputChange} />);
            // After Input
            retElems.push(<span style={style} key={`${textCounter}-after`}>{entry.text.substr(lesson.input.endIndex - startIndex, entry.text.length)}</span>)
        } else {
            retElems.push(<span style={style} key={textCounter}>{entry.text}</span>)
        }
        
        textCounter += entry.text.length;

        return retElems;
    });

    return flatten(elems);
}