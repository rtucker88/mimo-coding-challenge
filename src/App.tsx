import * as React from 'react';
import { BrowserRouter as Router, Redirect, Route, RouteComponentProps, Switch } from 'react-router-dom';

import { ILessonCompletion } from './services/database';
import { getLessons, ILesson } from './services/lesson';

import Done from './Done';
import Lesson from './Lesson';

interface IAppProps {
  saveLesson: (lesson: ILessonCompletion) => Promise<any>;
}

interface IAppState {
  currentLesson: number | 'DONE';
  isLoading: boolean;
  lessons: ILesson[];
}

class App extends React.Component<IAppProps, IAppState> {
  // Keep this out of state since it's in no-way tied to rendering or updating
  private currentLessonStart: string;

  constructor(props: IAppProps) {
    super(props);

    this.state = {
      currentLesson: 0,
      isLoading: true,
      lessons: [],
    };
  }

  public async componentDidMount() {
    // Assuming we always start at the first lesson and aren't trying to "continue from the last lesson"
    const lessons = await getLessons();
  
    this.setState({
      currentLesson: lessons[0].id,
      isLoading: false,
      lessons,
    });

    this.currentLessonStart = Date.now().toString();
  }

  public render() {
    const { isLoading, lessons } = this.state;

    return (
      <Router>
        <Switch>
          <Route path="/done" exact={true} component={Done} />
          <Route path="/:id" render={this.renderLesson} />
          { !isLoading ? <Redirect to={`/${lessons[0].id}`} /> : null } 
        </Switch>
      </Router>
    );
  }

  private renderLesson = ({ match }: RouteComponentProps<{ id: string }>) => {
    const { lessons } = this.state;
    const lesson = lessons.find(lsn => lsn.id === parseInt(match.params.id, 10));

    // This conditional deals with a sane default for the loading state
    return lesson ? (<Lesson lesson={lesson} key={lesson.id} nextLesson={this.getNextLesson(lesson.id)} completeLesson={this.completeLesson} />) : null;
  }

  private getNextLesson = (id: number) => {
    const { lessons } = this.state;
    const currentIndex = lessons.findIndex(lesson => lesson.id === id);

    return currentIndex !== lessons.length - 1 ? lessons[currentIndex + 1].id : 'DONE'
  };

  private completeLesson = async () => {
    // Unhandled error case here, if an error would result when adding to the DB
    await this.props.saveLesson({
      end: Date.now().toString(),
      id: this.state.currentLesson as number,
      start: this.currentLessonStart,
    });
    this.currentLessonStart = Date.now().toString();

    this.setState((prevState) => ({
      currentLesson: this.getNextLesson(prevState.currentLesson as number),
    }));
  }
}

export default App;
