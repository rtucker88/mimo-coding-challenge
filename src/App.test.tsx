import { mount } from 'enzyme';
import * as lolex from 'lolex';
import * as React from 'react';

import App from './App';
import Lesson from './Lesson';

jest.mock('./services/lesson');

import { lessons as mockLessons } from './services/__mocks__/lesson';

describe('App', () => {
  const saveLessonMock = jest.fn();

  afterEach(() => {
    saveLessonMock.mockReset();
  });

  it('renders correctly', () => {
    const wrapper = mount(<App saveLesson={saveLessonMock} />);
    expect(wrapper).toMatchSnapshot();
  });

  it('should set the initial state', (done) => {
    const wrapper = mount(<App saveLesson={saveLessonMock} />);
    expect(wrapper.state('isLoading')).toEqual(true);

    setImmediate(() => {
      const currentLesson = wrapper.state('currentLesson');
      const isLoading = wrapper.state('isLoading');
      const lessons = wrapper.state('lessons');

      expect(currentLesson).toEqual(mockLessons[0].id);
      expect(isLoading).toEqual(false);
      expect(lessons).toEqual(mockLessons);
      done();
    });
  });

  describe('lesson progression', () => {
    it('should progress to the next lesson', (done) => {
      const wrapper = mount(<App saveLesson={saveLessonMock} />);
      
      setImmediate(() => {
        wrapper.update();
        const lesson = wrapper.find(Lesson);
        lesson.prop('completeLesson')();

        setImmediate(() => {
          wrapper.update();
          expect(wrapper.state('currentLesson')).toEqual(1);
          done();
        });
      });
    });

    it('should complete the lessons', (done) => {
      const wrapper = mount(<App saveLesson={saveLessonMock} />);
      
      setImmediate(() => {
        wrapper.update();
        const lesson = wrapper.find(Lesson);
        wrapper.setState({ currentLesson: 1 });
        lesson.prop('completeLesson')();

        setImmediate(() => {
          wrapper.update();
          expect(wrapper.state('currentLesson')).toEqual('DONE');
          done();
        });
      });
    });
  });

  it('should call saveLesson', (done) => {
    const clock = lolex.install({ toFake: ['Date']});

    const wrapper = mount(<App saveLesson={saveLessonMock} />);
    
    setImmediate(() => {
      wrapper.update();
      const lesson = wrapper.find(Lesson);
      clock.setSystemTime(15);
      lesson.prop('completeLesson')();

      setImmediate(() => {
        wrapper.update();
        expect(wrapper.state('currentLesson')).toEqual(1);

        expect(saveLessonMock.mock.calls[0][0]).toEqual({
          end: "15",
          id: 0,
          start: "0",
        });

        clock.uninstall();
        done();
      });
    });
  });
});
