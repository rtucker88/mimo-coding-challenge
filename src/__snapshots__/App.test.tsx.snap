// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App renders correctly 1`] = `
<App
  saveLesson={[MockFunction]}
>
  <BrowserRouter>
    <Router
      history={
        Object {
          "action": "POP",
          "block": [Function],
          "createHref": [Function],
          "go": [Function],
          "goBack": [Function],
          "goForward": [Function],
          "length": 1,
          "listen": [Function],
          "location": Object {
            "hash": "",
            "pathname": "/",
            "search": "",
            "state": undefined,
          },
          "push": [Function],
          "replace": [Function],
        }
      }
    >
      <Switch />
    </Router>
  </BrowserRouter>
</App>
`;
